packages/firestore/lib/modular/snapshot.d.ts

```diff
@@ -0,0 +1,203 @@
+import { FirebaseFirestoreTypes } from '../index';
+
+import DocumentReference = FirebaseFirestoreTypes.DocumentReference;
+import DocumentSnapshot = FirebaseFirestoreTypes.DocumentSnapshot;
+import SnapshotListenOptions = FirebaseFirestoreTypes.SnapshotListenOptions;
+import QuerySnapshot = FirebaseFirestoreTypes.QuerySnapshot;
+import Query = FirebaseFirestoreTypes.Query;
+
+export type Unsubscribe = () => void;
+export type FirestoreError = Error;
+
+/**
+ * Attaches a listener for `DocumentSnapshot` events. You may either pass
+ * individual `onNext` and `onError` callbacks or pass a single observer
+ * object with `next` and `error` callbacks.
+ *
+ * NOTE: Although an `onCompletion` callback can be provided, it will
+ * never be called because the snapshot stream is never-ending.
```

Comment:
Does `onCompletion()` not get called when the returned `unsubscribe()` is called?

Reply:
interesting question! Honestly I'm not sure - you'd have to chase in to the native code I think and see what events are passed native<->javascript

Alternatively, if nothing is provided then it does seem reasonable to call onCompletion when unsubscribe happens, as by definition that would mean there will never be more events from that subscription...

Reply:
Hmmm....not promising for Android: https://github.com/search?q=repo%3Afirebase%2Ffirebase-android-sdk%20oncompletion&type=code

Though apparently it is a bigger deal for iOS? https://github.com/search?q=repo%3Afirebase%2Ffirebase-ios-sdk+oncompletion&type=code

Reply:
I just looked more closely at the firebase-js-sdk documentation, which we generally treat as our authority here with regard to what behavior contract to meet for APIs, and it's pretty clear (and is the source of that comment I think):

https://firebase.google.com/docs/reference/js/firestore_.md#onsnapshot

> NOTE: Although an onCompletion callback can be provided, it will never be called because the snapshot stream is never-ending.

I do not think we can call the onCompletion callback *and* meet the behavior contract of onSnapshot then, so I don't think this is a productive path to go down with regard to investigation. Sorry I started down it with you, if that meant you spent any significant amount of time on it